### ISSUE TITLE
JavaClassElement.getTypeArguments(Class) gives wrong results for parameterized types

### ISSUE THREAD

**Issue by yawkat** [micronaut-projects/micronaut-core#5798]
`JavaClassElement.getTypeArguments(Class)` uses the type parameters from the `TypeElement` to resolve type params of super parameters, but `TypeElement` represents a type declaration (e.g. `interface List<E> extends Collection<E>`) and not a type use (`List<String>`). So if you do `.getTypeArguments(Collection.class)` on a `List<String>`, it'll return `{"E" -> java.lang.Object}`.

Looking at how `getTypeArguments(Class)` used elsewhere, it's only used in contexts where the `ClassElement` actually represents a type declaration, not on a type use as the declared type of a field.

**Comment by graemerocher** [comment]
I don't believe that `TypeElement` represents a type declaration. According to the java it "Represents a class or interface program element"

`DeclaredType` represents a type declaration.

**Comment by graemerocher** [comment]
Having said that for issue reports you need to follow the template and maybe explain the use case and where this is manifesting as an issue with an example

**Comment by yawkat** [comment]
Yes, it seems like `TypeElement` lacking this info is the reason for this bug. When `JavaClassElement` is used for a parameterized type, i.e. when `genericTypeInfo` is non-empty, this additional info is ignored by `getTypeArguments`. Test case (place in `ClassElementSpec`):

```groovy
    void "test resolve generic method return type using getTypeArguments"() {
        buildBeanDefinition('test.TestController', '''
package test;

import io.micronaut.http.annotation.*;

@Controller("/test")
public class Test {
    @Get("/getMethod")
    List<String> list() { return null; }
}

''')
        expect:
        AllElementsVisitor.VISITED_METHOD_ELEMENTS.size() == 1
        AllElementsVisitor.VISITED_METHOD_ELEMENTS[0].getGenericReturnType().getTypeArguments(Iterable.class)["T"].name == String.name
    }
```

This returns `java.lang.Object` instead of `java.lang.String`.

**Comment by graemerocher** [comment]
Triage: this appears completed in the current annotation-processing API. ClassElement.getTypeArguments(Class) now resolves through getAllTypeArguments(), and JavaClassElement resolves arguments from the parameterized TypeMirror; current generic hierarchy work also has regression coverage for intermediate super-type bindings. I could not justify keeping this old report open without a current failing test. Closing as completed; reopen with the exact current processor test if Collection<E> still resolves to Object for List<String>.
